### PR TITLE
Add unique SEO content to news coin index #830067084883312

### PR DESCRIPTION
**Pull request by jessehon**
https://app.asana.com/0/683768070860536/830067084883312

image

### CODE REVIEW

File: app/models/coin.rb [hunk 1]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
+        Trading volume for #{self.name} over the last 24 hours is
+        #{(self.market_info["24h_volume_usd"])}.
+      ]
+    end
+
+    recent_news_count = self.news_items
+      .where('feed_item_published_at >= ?', 7.days.ago)
+      .count()
+    formatted_recent_news_count = recent_news_count == 0 ? 'no' : recent_news_count
+    if recent_news_count > 1
```
Comment (hanchang) [inline app/models/coin.rb]: Can probably simplify this to:

```
if recent_news_count == 1
  result << %W[ There has been 1 news story on #{self.name} over the last 7 days. ]
else
  result << %W[ There have been #{recent_news_count} news stories on #{self.name} over the last 7 days. ]
end
```

File: app/models/coin.rb [hunk 2]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
+      .first()
+  end
+
+  def summary
+    result = [];
+
+    if self.ranking && self.market_info["market_cap_usd"]
+      result << %W[
+        #{self.name} (#{self.symbol}) is currently the ##{self.ranking} cryptocurrency by market cap
+        at #{(self.market_info["market_cap_usd"])}.
```
Comment (hanchang) [inline app/models/coin.rb]: Let's be explicit and add "USD", i.e.

```
cryptocurrency by market cap at #{(self.market_info["market_cap_usd"])} USD
```

File: app/models/coin.rb [hunk 3]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
+      ]
+    end
+
+    if self.market_info["24h_volume_usd"]
+      result << %W[
+        Trading volume for #{self.name} over the last 24 hours is
+        #{(self.market_info["24h_volume_usd"])}.
```
Comment (hanchang) [inline app/models/coin.rb]: Same explicitness

File: app/models/coin.rb [hunk 4]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
   end
 
+  def most_common_news_category
+    NewsCategory
+      .joins(news_items: :news_coin_mentions)
+      .where(news_coin_mentions: { coin_id: self })
+      .group(:id)
+      .order('COUNT(DISTINCT news_items.id) DESC')
+      .limit(1)
+      .first()
```
Comment (hanchang) [inline app/models/coin.rb]: Let's stick with Ruby conventions of no parens for function calls without parameters.

File: app/models/coin.rb [hunk 5]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
+  end
+
+  def most_common_feed_source
+    FeedSource
+      .joins(news_items: :news_coin_mentions)
+      .where(news_coin_mentions: { coin_id: self })
+      .group(:id)
+      .order('COUNT(DISTINCT news_items.id) DESC')
+      .limit(1)
+      .first()
```
Comment (hanchang) [inline app/models/coin.rb]: Let's stick with Ruby conventions of no parens for function calls without parameters.

File: app/models/coin.rb [hunk 6]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
+    if self.market_info["24h_volume_usd"]
+      result << %W[
+        Trading volume for #{self.name} over the last 24 hours is
+        #{(self.market_info["24h_volume_usd"])}.
+      ]
+    end
+
+    recent_news_count = self.news_items
+      .where('feed_item_published_at >= ?', 7.days.ago)
+      .count()
```
Comment (hanchang) [inline app/models/coin.rb]: Let's stick with Ruby conventions of no parens for function calls without parameters.

File: app/models/coin.rb [hunk 7]
```diff
@@ -52,6 +52,72 @@ def self.symbols
...
+
+    most_common_feed_source_name = self.most_common_feed_source.name
+    most_common_news_category_name = self.most_common_news_category.name
+    result << %W[
+      The most common news source covering #{self.name} is #{most_common_feed_source_name} and the
+      most common news category is #{most_common_news_category_name}.
+    ]
+
+    # Merge the result into a single line string
+    result.flatten().join(' ')
```
Comment (hanchang) [inline app/models/coin.rb]: Let's stick with Ruby conventions of no parens for function calls without parameters.

File: app/models/coin.rb (hunk 1 above)
Comment (hanchang) [reply app/models/coin.rb]: Nevermind, yours is better!

**Comment by jessehon** [comment]
Query performance is pretty bad so not sure we want to merge this. It does only affect the coin detail pages though

image

**Comment by hanchang** [comment]
Ouch, thanks for the heads - let's work on improving the query performance then, thanks!

**Comment by jessehon** [comment]
> Ouch, thanks for the heads - let's work on improving the query performance then, thanks!

Doesn't seem to be a big performance issue anymore

**Comment by hanchang** [comment]
@jessehon Feel free to merge whenever you're ready, LGTM still 👌

**Pull request merged**